broadcastclients/broadcastclients.go

```diff
@@ -72,12 +126,117 @@ func (bcs *BroadcastClients) adjustCount(delta int32) {
+					continue
+				}
+				if _, ok := recentFeedItemsOld[msg.SequenceNumber]; ok {
+					continue
+				}
+				recentFeedItemsNew[msg.SequenceNumber] = time.Now()
+				if err := bcs.primaryRouter.forwardTxStreamer.AddBroadcastMessages([]*broadcaster.BroadcastFeedMessage{&msg}); err != nil {
+					log.Error("Error routing message from Primary Sequencer Feeds", "err", err)
+				}
+			case cs := <-bcs.primaryRouter.confirmedSequenceNumberChan:
+				startSecondaryFeedTimer.Reset(MAX_FEED_INACTIVE_TIME)
+				primaryFeedIsDownTimer.Reset(MAX_FEED_INACTIVE_TIME)
+				if cs == lastConfirmed {
+					continue
+				}
+				lastConfirmed = cs
+				bcs.primaryRouter.forwardConfirmationChan <- cs
+
+			// Secondary Feeds
+			case msg := <-bcs.secondaryRouter.messageChan:
+				startSecondaryFeedTimer.Reset(MAX_FEED_INACTIVE_TIME)
+				if _, ok := recentFeedItemsNew[msg.SequenceNumber]; ok {
+					continue
+				}
+				if _, ok := recentFeedItemsOld[msg.SequenceNumber]; ok {
+					continue
+				}
+				recentFeedItemsNew[msg.SequenceNumber] = time.Now()
+				if err := bcs.secondaryRouter.forwardTxStreamer.AddBroadcastMessages([]*broadcaster.BroadcastFeedMessage{&msg}); err != nil {
+					log.Error("Error routing message from Secondary Sequencer Feeds", "err", err)
+				}
+			case cs := <-bcs.secondaryRouter.confirmedSequenceNumberChan:
+				startSecondaryFeedTimer.Reset(MAX_FEED_INACTIVE_TIME)
+				if cs == lastConfirmed {
+					continue
+				}
+				lastConfirmed = cs
+				bcs.secondaryRouter.forwardConfirmationChan <- cs
+
+			// Cycle buckets to get rid of old entries
+			case <-recentFeedItemsCleanup.C:
+				recentFeedItemsOld = recentFeedItemsNew
+				recentFeedItemsNew = make(map[arbutil.MessageIndex]time.Time, RECENT_FEED_INITIAL_MAP_SIZE)
+
+			// failed to get messages from both primary and secondary feeds for ~5 seconds, start a new secondary feed
+			case <-startSecondaryFeedTimer.C:
+				bcs.StartSecondaryFeed(ctx)
+
+			// failed to get messages from primary feed for ~5 seconds, reset the timer responsible for stopping a secondary
+			case <-primaryFeedIsDownTimer.C:
+				stopSecondaryFeedTimer.Reset(PRIMARY_FEED_UPTIME)
+
+			// primary feeds have been up and running for PRIMARY_FEED_UPTIME=10 mins without a failure, stop the recently started secondary feed
+			case <-stopSecondaryFeedTimer.C:
+				bcs.StopSecondaryFeed(ctx)
+			}
+		}
+	})
 }
+
+func (bcs *BroadcastClients) StartSecondaryFeed(ctx context.Context) {
+	if bcs.numOfStartedSecondary < len(bcs.secondaryClients) {
+		client := bcs.secondaryClients[bcs.numOfStartedSecondary]
+		bcs.numOfStartedSecondary += 1
+		client.Start(ctx)
+	} else {
+		log.Warn("failed to start a new secondary feed all available secondary feeds were started")
+	}
+}
+func (bcs *BroadcastClients) StopSecondaryFeed(ctx context.Context) {
+	if bcs.numOfStartedSecondary > 0 {
+		bcs.numOfStartedSecondary -= 1
+		client := bcs.secondaryClients[bcs.numOfStartedSecondary]
+		client.StopAndWait()
+	}
+}
+
```

Comment:
nit: Suggest making these private since they're not thread safe.

Reply:
ahh, agreed!